include/ttmlir/Dialect/TTIR/Transforms/Passes.td

```diff
@@ -108,6 +108,14 @@ def TTIRAllocate: Pass<"ttir-allocate", "::mlir::ModuleOp"> {
   }];
 }
 
+def TTIRDecompose: Pass<"ttir-decompose", "::mlir::ModuleOp"> {
+  let summary = "Decompose ops to more simpler ops";
+  let description = [{
+    Go through the ops, set sharding specs for each op based on sharding analysis,
+    by updating layout attribute of each op.
+  }];
+}
+
```

Comment:
I wonder if instead of doing a decompose pass, we use MLIR's canonicalizer framework, see https://mlir.llvm.org/docs/Canonicalization/.

It essentially boils down to declaring `let hasCanonicalizer = 1;` in the op's tablegen (in this case `IndexOp` above) and then implementing (see `ArrayCoorOp` as an example in LLVM):
https://github.com/llvm/llvm-project/blob/c3aa8b7dd62065714256ad439afeab6fb2b1b89f/flang/lib/Optimizer/Dialect/FIROps.cpp#L810
```c++
 void fir::ArrayCoorOp::getCanonicalizationPatterns(      
      mlir::RewritePatternSet &patterns, mlir::MLIRContext *context) {  
    // TODO: !fir.shape<1> operand may be removed from array_coor always.
    patterns.add<SimplifyArrayCoorOp>(context);
  }
```

@sdjordjevicTT

Reply:
@sdjordjevicTT, @svuckovicTT, I think this boils down to how we want to handle decomposition in TTIR, we could try and get a read on ways to do this at the MLIR conference.

I think the canonicalizer might always run which means that it'll always present a decomp'd graph, it's unclear if this is desirable.  I could see some cases where you might want to keep the higher level op around, not sure.

Reply:
Can you list some of the benefits we gain by doing decomposition in the conversion pass?

Reply:
@jserbedzijaTT a big one is that we get to define a set of legal ops within the dialect conversion api. In the "regular optimization pass" path, there's no such feature, so we'd have to hack something up manually.

To a lesser extent, we also get type conversion for free, and even though it doesn't look like we'll need it for now, it's nice to know it'll be there of we do.

Reply:
I like the TTIRToTTIR dialect conversion idea, let's do it this way